modules/cloud-function/main.tf

```diff
@@ -39,6 +39,16 @@ locals {
     )
     : var.service_account
   )
+  trigger_service_account_email = (
```

Comment:
The idea is that is either null or an SA email. Let's dissect the expression:
```hcl
    try(var.trigger_config.v2.service_account_create, null) == null
    ? false
    : var.trigger_config.v2.service_account_create ? (
      length(google_service_account.trigger_service_account) > 0
      ? google_service_account.trigger_service_account[0].email
      : null
    )
    : try(var.trigger_config.v2.service_account_email, null)
```

So first part:
```hcl
    try(var.trigger_config.v2.service_account_create, null) == null
    ? false
    : var.trigger_config.v2.service_account_create 
```
gives either false or value of `var.trigger_config.v2.service_account_create` when it is not null.

Then the result of above expression is used to choose whether:
````hcl
above_expression ? (
      length(google_service_account.trigger_service_account) > 0
      ? google_service_account.trigger_service_account[0].email
      : null
    )
    : try(var.trigger_config.v2.service_account_email, null)
```
Return email of service account that was created by resource or email that was provided in `var.trigger_config.v2.service_account_email`.

Hopefully I wrote what I meant. When do you think it returns `false`?

Regrettably adding extra brackets doesn't help with tf formating:
```hcl
  trigger_service_account_email = (
    (
      try(var.trigger_config.v2.service_account_create, null) == null
      ? false
      : var.trigger_config.v2.service_account_create
      ) ? (length(google_service_account.trigger_service_account) > 0
      ? google_service_account.trigger_service_account[0].email
      : null
    )
    : try(var.trigger_config.v2.service_account_email, null)
  )
``` 
I'd love the line `) ? (length(google_service_account.trigger_service_account) > 0` to be one-step less indented than  the surrounding ones.

Reply:
> When do you think it returns false?

The very first branch returns a false:
```hcl
  trigger_service_account_email = (
    (
      try(var.trigger_config.v2.service_account_create, null) == null
      ? false # <--- here
```


I would format it like this:
```hcl
trigger_service_account_email = (
  try(var.trigger_config.v2.service_account_create, null) == null
  ? false
  : (
    var.trigger_config.v2.service_account_create
    ? (
      length(google_service_account.trigger_service_account) > 0
      ? google_service_account.trigger_service_account[0].email
      : null
    )
    : try(var.trigger_config.v2.service_account_email, null)
  )
)
```
But that's besides the point. You have 3 nested conditions that are really hard to follow.

I'm going to throw a few (untested) ideas for you to explore:

```hcl
locals{
  # use try for the inner conditional
  trigger_service_account_email1 = (
    try(var.trigger_config.v2.service_account_create, null) == null
    ? false
    : (
      var.trigger_config.v2.service_account_create
      ? try(google_service_account.trigger_service_account[0].email, null)
      : try(var.trigger_config.v2.service_account_email, null)
    )
  )

  # use try to check create flag
  trigger_service_account_email2 = (
    try(var.trigger_config.v2.service_account_create, false)
    ? try(google_service_account.trigger_service_account[0].email, null)
    : try(var.trigger_config.v2.service_account_email, null)
  )

  # just use try
  trigger_service_account_email3 = try(
    google_service_account.trigger_service_account[0].email,
    var.trigger_config.v2.service_account_email,
    null
  )
}
```

Reply:
Just for the record, the whole expression:
```hcl
    try(var.trigger_config.v2.service_account_create, null) == null
    ? false
    : var.trigger_config.v2.service_account_create 
```
Returns `true` or `false`. This is later used in following expression. I agree that this is far to complex to follow. I came up with this simplification:
```hcl
  trigger_service_account_email = (
    coalesce(try(var.trigger_config.v2.service_account_create, false), false)
    ? google_service_account.trigger_service_account[0].email
    : null
  )
```

The part with checking whether the service account was created or not I took from what is already there:
```hcl
  service_account_email = (
    var.service_account_create
    ? (
      length(google_service_account.service_account) > 0
      ? google_service_account.service_account[0].email
      : null
    )
    : var.service_account
  )
```
It's kind of really defensive programming as `var.service_account_create` is used to control, whether service account is created at all. I guess this could too can be simplified to
``hcl
  service_account_email = (
    var.service_account_create
    ? google_service_account.service_account[0].email
    : var.service_account
  )
```
or please help me understand what I'm missing there .

Reply:
My whole point is that the original expression is unnecessarily complex. If you can simplify it without compromising the logic, you should.

Reply:
Done for `trigger_service_account` in 0cb1f6ae3c822bfe7dd2d37d30861178eac1ef61 and for `service_account_email` in 78d20b1f16307cd3c5d5bbdd7c8f9bfd6bbf7f7d